libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,467 @@
+
+    // @Param: OUTPUT
+    // @DisplayName: Output database
+    // @Description: Enables streaming of the entire database as ADSB objects. This requires a high-bandwidth connection to your GCS
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Standard
+    AP_GROUPINFO("OUTPUT", 3, AP_OADatabase, _output_as_adsb, 0),
+
+AP_GROUPEND
+};
+
+AP_OADatabase::AP_OADatabase()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_OADatabase::init()
+{
+    if (_initialised || _unable_to_allocate_memory || _database.size_param < (int32_t)_database.size_min) {
+        return;
+    }
+
+    init_database();
+    init_queue();
+
+    if (_queue.items != nullptr && _database.items != nullptr) {
+        ::printf("DB initialised. Sizes queue:%d, db:%d\n", _queue.size, _database.size);
+        _initialised = true;
+    }
+}
+
+void AP_OADatabase::init_queue()
+{
+    if (_queue.items != nullptr) {
+        return;
+    }
+    _queue.size = constrain_int32(_queue.size, _queue.size_min, _queue.size_max);
+    ::printf("DB::init_queue() %d\n", _queue.size);
+
+    _queue.items = new ObjectBuffer<OA_DbItem>(_queue.size);
+    if (_queue.items == nullptr) {
+        if (_queue.size <= _queue.size_min) {
+            // We're unable to allocate memory with the smallest allowed size. hard failure, do not retry.
+            ::printf("DB failed to allocate queue of size %d - not retrying\n", _queue.size);
+            _unable_to_allocate_memory = true;
+        } else {
+            // failed to allocate, try a little smaller next time
+            ::printf("DB failed to allocate queue of size %d\n", _queue.size);
+            _queue.size *= _queue.size_shrink_rate;
+        }
+    }
+}
+
+void AP_OADatabase::init_database()
+{
+    if (_database.items != nullptr || _database.size_param < (int32_t)_database.size_min) {
+        return;
+    }
+
+    _database.size = _database.size_param;
+    _database.count = 0;
+    ::printf("DB::init_database() %d\n", _database.size);
+
+    _database.items = (OA_DbItem *)calloc(_database.size, sizeof(OA_DbItem));
+    if (_database.items == nullptr) {
+        if (_database.size <= _database.size_min) {
+            // We're unable to allocate memory with the smallest allowed size. hard failure, do not retry.
+            ::printf("DB failed to allocate database of size %d - not retrying\n", _database.size);
+            _unable_to_allocate_memory = true;
+        } else {
+            // failed to allocate, try a little smaller next time
+            ::printf("DB failed to allocate database of size %d\n", _database.size);
+            _database.size_param.set_and_notify(_database.size_param * 0.9);
+        }
+    }
+    _database.size_param_last = _database.size_param;
+}
+
+void AP_OADatabase::deinit_database()
+{
+    _initialised = false;
+    _database.size_param_last = _database.size_param;
+
+    _database.size = 0;
+    _database.count = 0;
```

Comment:
Should reset the last hit index.

Reply:
last hit index removed so marking as resolved